### PR TITLE
Documentation - CI/CD Tutorial End to End with Skaffold

### PR DESCRIPTION
**Pull request by tvvignesh**
Fixes: #346 

**Description**
This PR add documentation for CI/CD end to end with skaffold

### CODE REVIEW

**Comment by tvvignesh** [comment]
@googlebot I signed it!

**Comment by gsquared94** [comment]
Please take a look @viglesiasce

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+**Step 3: Choosing the image and tag strategy**
+
+Skaffold automatically tags images based on different strategies as documented here: [https://skaffold.dev/docs/pipeline-stages/taggers/](https://skaffold.dev/docs/pipeline-stages/taggers/) and you might want to choose the mechanism which is right for you. By default, it uses the git sha if unspecified.
+
+Also, if you are using helm chart as your deployment method, you might want to use the helm image strategy if you would like to follow helm specific conventions and skaffold will pass in all the details for you.
+
+**Step 4: Plugins (optional)**
+
+Another interesting thing is that skaffold also supports plugins like Helm Secrets ([https://github.com/zendesk/helm-secrets/](https://github.com/zendesk/helm-secrets/)). So, if you would like to deploy secrets that have been encrypted using KMS or any other mechanism supported by SOPS ([https://github.com/mozilla/sops](https://github.com/mozilla/sops)) you can actually set **useHelmSecrets** option to true and skaffold will handle everything automatically for you.
```
Comment (MarlonGamez) [inline docs/content/en/docs/tutorials/ci_cd.md:44]:
```suggestion
Skaffold also supports plugins like Helm Secrets ([https://github.com/zendesk/helm-secrets/](https://github.com/zendesk/helm-secrets/)). So, if you would like to deploy secrets that have been encrypted using KMS or any other mechanism supported by SOPS ([https://github.com/mozilla/sops](https://github.com/mozilla/sops)) you can actually set **useHelmSecrets** option to true and skaffold will handle everything automatically for you.
```

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+**Step 5: The skaffold.yaml file**
+
+Depending on the way you would like to build, deploy, your skaffold yaml file will actually change and it's quite easy to set up one. It is recommended that you set up skaffold before you setup your CI pipeline since the CI pipeline can just call skaffold to do all the builds and deployments and you can do the same locally as well.
```
Comment (MarlonGamez) [inline docs/content/en/docs/tutorials/ci_cd.md:48]:
```suggestion
Your skaffold.yaml file will change depending on the way you would like to build and deploy your project. It is recommended that you set up skaffold before you setup your CI pipeline since the CI pipeline can just call skaffold to do all the builds and deployments, and you can do the same locally as well.
```

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+                        NPM_TOKEN: '{{.NPM_TOKEN}}'
+      deploy:
+          statusCheckDeadlineSeconds: 600
+          kubectl:
+              manifests:
+                  - k8/dev/*.yml
+```
+
+
+**Please refer the yaml reference for all that you can specify in skaffold.yaml file**
```
Comment (MarlonGamez) [inline docs/content/en/docs/tutorials/ci_cd.md:120]:
```suggestion
**Please refer to the yaml reference for all that you can specify in skaffold.yaml file**
```

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+
+## **Step 6: Development & Testing locally**
```
Comment (MarlonGamez) [inline docs/content/en/docs/tutorials/ci_cd.md:123]: The labels for steps 1-5 use emphasis, but steps 6-9 use emphasis and are also headers. This styling should be consistent throughout.

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+*   And we trigger this only on changes to master branch
+
+So, ultimately our Gitlab CI pipeline does not know anything about the helm charts or build strategy or any deployment methods and relies completely on skaffold to do the job for us.
+
+This works equally well irrespective of whether you choose the Docker Executor ([https://docs.gitlab.com/runner/executors/docker.html](https://docs.gitlab.com/runner/executors/docker.html)) or the Kubernetes Executor ([https://docs.gitlab.com/runner/executors/kubernetes.html](https://docs.gitlab.com/runner/executors/kubernetes.html)) and would require little to no changes.
+
+**NOTE:** If you are running Skaffold using Docker, make sure that you are running the runner with appropriate permissions or pod security policies since running DIND requires access to the Docker socket and being in privileged mode which might actually tend to be insecure. An alternative can be to use kaniko or custom builders like Buildah for building the image.
+
+
+## **Step 9: Using DIND Service (Optional)**
```
Comment (MarlonGamez) [inline docs/content/en/docs/tutorials/ci_cd.md:187]: Numbering for steps 8 and 9 should be flipped

**Review by MarlonGamez — CHANGES_REQUESTED** [review]
Thanks for submitting this PR @tvvignesh! I've taken a first pass at it, mostly just commenting on nits related to readability and language, not so much on content. Let me know what you think of my suggestions :)

**Comment by tvvignesh** [comment]
@MarlonGamez Thanks a lot for your suggestions. Have made all the necessary changes. Let me know if there is anything else to be done.

**Comment by tejal29** [comment]
Thanks @tvvignesh And @MarlonGamez for taking this forward!

**Comment by MarlonGamez** [comment]
@tvvignesh thanks a bunch for making the changes :) One thing I would say is that I think the step labels should be h2s (like step 6 was previously), just to make a bit more clear where each step starts

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+*   Logging in to our gcp account using the service account key
+*   Specifying the project, zone, cluster to deploy our application in.
+*   We are using DIND (Docker In Docker) to facilitate our builds (more about that below)
+*   Then we just call the skaffold run in the pipeline passing flags we want to pass at runtime (in our case our cluster context, env vars, namespace and some labels)
+*   And we trigger this only on changes to master branch
+
+So, ultimately our Gitlab CI pipeline does not know anything about the helm charts or build strategy or any deployment methods and relies completely on skaffold to do the job for us.
+
+This works equally well irrespective of whether you choose the Docker Executor ([https://docs.gitlab.com/runner/executors/docker.html](https://docs.gitlab.com/runner/executors/docker.html)) or the Kubernetes Executor ([https://docs.gitlab.com/runner/executors/kubernetes.html](https://docs.gitlab.com/runner/executors/kubernetes.html)) and would require little to no changes.
```
Comment (briandealwis) [inline docs/content/en/docs/tutorials/ci_cd.md:182]: This assumes some knowledge about Gitlab.  I *think* you're using the Kubernetes executor here (since you have a deployment yaml below, which doesn't look related to your application).  Why did you choose this if you could use the Docker Executor?

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+Here, we are:
+
+
+
+*   Logging into the docker registry specifying the Registry URL, User ID and password (You can choose to authenticate using service tokens as well or other auth providers as well)
+*   Logging in to our gcp account using the service account key
+*   Specifying the project, zone, cluster to deploy our application in.
+*   We are using DIND (Docker In Docker) to facilitate our builds (more about that below)
+*   Then we just call the skaffold run in the pipeline passing flags we want to pass at runtime (in our case our cluster context, env vars, namespace and some labels)
+*   And we trigger this only on changes to master branch
```
[the remark this answers is not in the document]
Comment (tvvignesh) [reply docs/content/en/docs/tutorials/ci_cd.md:178]: @briandealwis Thanks a lot for your suggestions.  Sure. Can change this. Yes, I use a private registry (Verdaccio) and that is what it authenticates with - it can be avoided, but I just put it here so that people would understand how to pass in args from gitlab to skaffold which then gets passed to the build args in Docker.

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+my-job:
+    image:
+        name: gcr.io/k8s-skaffold/skaffold:v1.15.0
+    tags:
+        - development
+    stage: my-stage
+    retry: 2
+    script:
+        - docker login $REG_DEV_URL -u $REG_DEV_UNAME -p $REG_DEV_PASSWORD
```
[the remark this answers is not in the document]
Comment (tvvignesh) [reply docs/content/en/docs/tutorials/ci_cd.md:151]: @briandealwis Actually, I used to have a private registry (Harbor) and this is how I used to authenticate to it. Since this is going to be a public tutorial, I replaced the image URL with gcr but then left this instruction out as such. This is for passing in user name, password and registry url in case you are either using docker hub or a private registry.

Should I change this to make it suitable to GCR? The process can be different for different registry provider, I thought this would be consistent. Let me know.

File: docs/content/en/docs/tutorials/ci_cd.md [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+You can find docs about all these workflows here: [https://skaffold.dev/docs/workflows/](https://skaffold.dev/docs/workflows/)
+
+Also, note that you can also set up file synchronization to enable faster development workflows. You can read more about that here: [https://skaffold.dev/docs/pipeline-stages/filesync/](https://skaffold.dev/docs/pipeline-stages/filesync/)
+
+
+## Step 7: Getting the CI pipeline setup
+
+Now that we have skaffold ready and working locally, the CI pipeline requires little to no effort since it will just call skaffold to do all the builds and deployments.
+
+Here is a sample pipeline using Gitlab CI
```
[the remark this answers is not in the document]
Comment (tvvignesh) [reply docs/content/en/docs/tutorials/ci_cd.md:136]: Hmm.. I understand your point. Since I work with Gitlab for my workflow, I went with this tutorial keeping Gitlab in mind and I hope other CI solutions are similar as well (I guess I may need some external opinion on this). I can add some explanation to this to add more clarity. Regarding REG_DEV_URL, I have mentioned ^^ I had actually replaced the private registry url with gcr since this is a public tutorial. and left REG_DEV_URL as is without removing it. I will replace all instances to suit GCR if that would be fine.

File: docs/content/en/docs/tutorials/ci_cd.md (hunk 6 above)
Comment (tvvignesh) [reply docs/content/en/docs/tutorials/ci_cd.md:182]: Docker executor does not run in the Kubernetes cluster, requires access to the Docker engine and also does not allow for things like scaling up/down runners as pods in the cluster all of which is supported in Kubernetes executor. While all the instructions mentioned in this tutorial will work in both runners, Kubernetes executor has its own set of advantages. Should I elaborate about it here?

Comment (briandealwis) [reply docs/content/en/docs/tutorials/ci_cd.md:182]: I think the example should use the simplest approach (which I'm guessing is the Docker executor).  You could include a pointer to the Kubernetes executor and briefly summarize some of its advantages.

**Comment by tvvignesh** [comment]
I will spend sometime next week on this and try to do a rewrite taking into account @briandealwis 's suggestions. Please leave all your suggestions here. Thanks.

**Comment by MarlonGamez** [comment]
hey @tvvignesh, have you had the chance to work on this a bit? No rush, just curious :) thanks again for working on this!

**Comment by tvvignesh** [comment]
> hey @tvvignesh, have you had the chance to work on this a bit? No rush, just curious :) thanks again for working on this!

@briandealwis @MarlonGamez Hi. Pushed the changes. Have tried to address all the feedback as given above.

Hope this is okay.

**Review by nkubala — APPROVED** [review]
in interest of not blocking this indefinitely, let's merge it for now and we can make edits later as we need them.

this would be a great addition to #5031 if it ever gets created.

thanks for all your work on this one @tvvignesh!

**Comment by tvvignesh** [comment]
@nkubala Forgot to inform this. I did not update `https://github.com/GoogleContainerTools/skaffold/blob/master/docs/content/en/docs/tutorials/_index.md` linking to this tutorial MD file deliberately since I wanted to leave it to you to decide where to link this in the website if needed.
